pkg/datapath/linux/ipsec/ipsec_linux.go

```diff
@@ -710,3 +676,83 @@ func StartKeyfileWatcher(ctx context.Context, keyfilePath string, nodediscovery
 
 	return nil
 }
+
+func doReclaimStaleKeys() {
+	ipSecKeysGlobalLock.Lock()
+	defer ipSecKeysGlobalLock.Unlock()
+
+	scopedLog := log.WithField("spi", ipSecCurrentKeySPI)
+
+	xfrmStateList, err := netlink.XfrmStateList(0)
+	if err != nil {
+		scopedLog.WithError(err).Warning("Failed to list XFRM states")
+		return
+	}
+
+	// First find when the current key was added
+	currentKeyCreatedAt := time.Time{}
+	for _, s := range xfrmStateList {
+		if s.Spi != int(ipSecCurrentKeySPI) {
+			continue
+		}
+
+		currentKeyCreatedAt = time.Unix(int64(s.Statistics.AddTime), 0)
+		break
+	}
+
+	// If the current key was added less then IPSecKeyDeleteDelay time ago,
+	// don't try to reclaim any stale key
+	if time.Now().Sub(currentKeyCreatedAt) < linux_defaults.IPsecKeyDeleteDelay {
+		return
+	}
+
+	// Otherwise loop again through the list of keys
+	for _, s := range xfrmStateList {
+		// If the key is the one currently in use, don't reclaim it
+		if s.Spi == int(ipSecCurrentKeySPI) {
+			continue
+		}
+
+		scopedLog := scopedLog.WithField("oldSPI", s.Spi)
+
+		// Otherwise delete the XFRM state for the key
+		scopedLog.Info("Deleting stale XFRM state")
+		if err := netlink.XfrmStateDel(&s); err != nil {
+			scopedLog.WithError(err).Warning("Deleting stale XFRM state failed")
+			continue
+		}
+
+		xfrmPolicyList, err := netlink.XfrmPolicyList(0)
+		if err != nil {
+			scopedLog.WithError(err).Warning("Failed to list XFRM policies")
+		}
+
+		// After the state, delete also any XFRM policy referencing the
+		// stale key
+		for _, p := range xfrmPolicyList {
+			if !ipSecPolicyForSPI(p, s.Spi) {
+				continue
+			}
+
+			if err := netlink.XfrmPolicyDel(&p); err != nil {
+				scopedLog.WithError(err).Warning("Deleting stale XFRM policy failed")
+			}
+		}
```

Comment:
You can move this second loop outside the first one if you switch `ipSecPolicyForSPI()` to a parser of the SPI (e.g., `getSPIFromXfrmPolicy()`) and just run through the list of XFRM policies, deleting those with a stale SPI, same as you're doing for XFRM states.